packages/parse5/lib/tokenizer/index.ts

```diff
@@ -220,17 +220,22 @@ export class Tokenizer {
 
     private consumedAfterSnapshot = -1;
 
+    private ctLoc: Location | null;
     private currentCharacterToken: CharacterToken | null = null;
     private currentToken: Token | null = null;
     private currentAttr: Attribute = { name: '', value: '' };
 
+    // NOTE: Doctypes tokens are created with several different offset. We keep track of the moment a doctype *might* start here.
+    private doctypeStartLoc: Location | null = null;
+
     private addLocationInfo;
     private onParseError;
 
     constructor(options: { sourceCodeLocationInfo?: boolean; onParseError?: ParserErrorHandler | null }) {
         this.addLocationInfo = !!options.sourceCodeLocationInfo;
         this.onParseError = options.onParseError ?? null;
         this.preprocessor = new Preprocessor(options);
+        this.ctLoc = this.getCurrentLocation(-1);
```

Comment:
not `0`?

Reply:
The doubly negated offset makes things a bit difficult here. An offset of `+1` means "start one character ago". `-1` means "start one character after our last seen one".

The preprocessor is initiated at position `-1`. We access the location here before we read anything, so have to give a negative offset of `-1` to get the position of the first character in the input. The same is done again in `_emitCurrentToken`.

I have played with flipping the sign of `getCurrentLocation`'s `offset`, but then other places are hard to read.